sqld/src/auth.rs

```diff
@@ -0,0 +1,236 @@
+use anyhow::{bail, Context as _, Result};
+
+/// Authentication that is required to access the server.
+#[derive(Default)]
+pub struct Auth {
+    /// When true, no authentication is required.
+    pub disabled: bool,
+    /// If `Some`, we accept HTTP basic auth if it matches this value.
+    pub http_basic: Option<String>,
+    /// If `Some`, we accept all JWTs signed by this key.
+    pub jwt_key: Option<jsonwebtoken::DecodingKey>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum AuthError {
+    #[error("The `Authorization` HTTP header is required but was not specified")]
+    HttpAuthHeaderMissing,
+    #[error("The `Authorization` HTTP header has invalid value")]
+    HttpAuthHeaderInvalid,
+    #[error("The authentication scheme in the `Authorization` HTTP header is not supported")]
+    HttpAuthHeaderUnsupportedScheme,
+    #[error("The `Basic` HTTP authentication scheme is not allowed")]
+    BasicNotAllowed,
+    #[error("The `Basic` HTTP authentication credentials were rejected")]
+    BasicRejected,
+    #[error("Authentication is required but no JWT was specified")]
+    JwtMissing,
+    #[error("Authentication using a JWT is not allowed")]
+    JwtNotAllowed,
+    #[error("The JWT is invalid")]
+    JwtInvalid,
+    #[error("The JWT has expired")]
+    JwtExpired,
+    #[error("The JWT is immature (not valid yet)")]
+    JwtImmature,
+    #[error("Authentication failed")]
+    Other,
+}
+
+/// A witness that the user has been authenticated.
+pub struct Authenticated(());
```

Comment:
question: what's the difference between `Authenticated` and `Authenticated(())`?

Reply:
`Authenticated(())` is a tuple struct with a single private field of type `()`. The point of this is that the value can only be created from this module, so it can act as a witness/proof that the user has been authenticated. This approach is inspired from rustls:
https://docs.rs/rustls/latest/rustls/manual/_01_impl_vulnerabilities/index.html#goto-fail

Reply:
interesting trick :)